cfi/lib/src/secmem.rs

```diff
@@ -0,0 +1,212 @@
+
+/// Values for a hardened boolean type.
+///
+/// The intention is that this is used instead of `<stdbool.h>`'s #bool, where a
+/// higher hamming distance is required between the truthy and the falsey value.
+///
+/// The values below were chosen at random, with some specific restrictions. They
+/// have a Hamming Distance of 8, and they are 11-bit values so they can be
+/// materialized with a single instruction on RISC-V. They are also specifically
+/// not the complement of each other.
+pub type HardenedBool = u32;
+pub const HARDENED_BOOL_TRUE: HardenedBool = 0x739;
+pub const HARDENED_BOOL_FALSE: HardenedBool = 0x1d4;
+
+struct RandomOrder {
+    state: u32,
+    max: u32,
+}
+
+/// Context for a random traversal order.
+///
+/// A "random traversal order" specifies a random order to walk through some
+/// buffer of length `n`, which is an important building block for
+/// constant-power code. Given `n`, the random order emits integers in the
+/// range `0..m`, where `m` is an implementation-defined, per-random-order
+/// value greater than `n`. The order is guaranteed to visit each integer in
+/// `0..n` at least once, but with some caveats:
+/// - Values greater than `n` may be returned.
+/// - The same value may be returned multiple times.
+///
+/// Users must be mindful of these constraints when using `RandomOrder`.
+/// These caveats are intended to allow for implementation flexibility, such as
+/// intentionally adding decoys to the sequence.
+impl RandomOrder {
+    /// Constructs a new, randomly-seeded traversal order,
+    /// running from `0` to at least `min_len`.
+    ///
+    /// This function does not take a seed as input; instead, the seed is
+    /// extracted, in some manner or another, from the hardware by this function.
+    ///
+    /// @param min_len The minimum length this traversal order must visit.
+    fn new(min_len: u32) -> RandomOrder {
+        RandomOrder {
+            state: 0,
+            max: min_len * 2,
+        }
+    }
+
+    /// Returns the length of the sequence represented by `ctx`.
+    ///
+    /// This value may be greater than `min_len` specified in
+    /// `random_order_init()`, but the sequence is guaranteed to contain every
+    /// integer in `0..min_len`.
+    ///
+    /// This value represents the number of times `random_order_advance()` may be
+    /// called.
+    ///
+    /// @param ctx The context to query.
+    /// @return The length of the sequence.
+    fn len(&self) -> u32 {
+        self.max
+    }
+
+    /// Returns the next element in the sequence represented by `ctx`.
+    ///
+    /// See `random_order_len()` for discovering how many times this function can
+    /// be called.
+    ///
+    /// @param ctx The context to advance.
+    /// @return The next value in the sequence.
+    fn advance(&mut self) -> u32 {
+        // TODO: The current implementation is just a skeleton, and currently just
+        // traverses from 0 to `min_len * 2`.
+        let s = self.state;
```

Comment:
The Gray code comment I made above is basically changing this to be `self.state ^ (self.state >> 1)`, and going to `min_len.next_power_of_two()` instead of `min_len * 2`. This added a few extra bytes of instruction code, which pushed it over the ROM limit, so I've removed it for now, and left this functionally equivalent to the not-at-all `random_walk` functions in OpenTitan.